### DISCUSSION TITLE
uv_timer_stop() does not stop and reset the timer

### DISCUSSION THREAD

**Discussion by nickyfow** [libuv/libuv#3831] [Q&A]
The `uv_timer_stop()` does not appear to stop and reset the timer. How can I stop and reset the timer correctly please? The timer duration is two seconds, and I start and stop it every second, so it should never reach the callback, but somehow it does:
```cpp
#include <iostream>
#include <thread>
#include <unistd.h> // usleep
#include <uv.h>
uv_loop_t *loop;
uv_timer_t timer_req;
float timerSeconds = 2.0f;
float sleepSeconds = 1.0f;

void run() {
  while (true) uv_run(loop, UV_RUN_DEFAULT);
}

void timerCallback(uv_timer_t* handle) {
  std::cout << "Callback" << std::endl;
}

int main() {
  loop = uv_default_loop();
  std::thread t(run);
  uv_timer_init(loop, &timer_req);
  while (true) {
    std::cout << "sleep" << std::endl;
    uv_timer_start(&timer_req, timerCallback, 1000*timerSeconds, 0);
    usleep(1000*1000*sleepSeconds);
    uv_timer_stop(&timer_req);
  };
  return 0;
}
```
**output on Debian Linux with libuv1-dev 1.24.1:**
```
$ g++ main.cc -o main `pkg-config --libs libuv`
$ ./main
sleep
sleep
sleep
sleep
sleep
Callback
sleep
^C
```
**Expected output:** The timer is set to call the callback after **two seconds** (`float timerSeconds = 2.0f`). But we stop (and restart) the timer after **every second** (`float sleepSeconds=1.0f`), so the timer should never run long enough for the callback to be executed and the output should show only the 'sleep' messages, instead of the 'Callback' message. That is, it should be:
```
$ g++ main.cc -o main `pkg-config --libs libuv`
$ ./main
sleep
sleep
[...]
```
I've tried:
- stopping the timer inside the callback
- reinitialising the timer every time: `using uv_timer_init()` inside the `while` loop, before calling `uv_timer_start()`

**Why I am doing this:** I'm writing a program to detect the long-press of a button (a short-press of a button triggers a different action). There is a callback to detect a **button-press**, and a **button-depress**. The button-press should start the timer, and button-depress should stop the timer. If the button has been pressed down long enough (two seconds), the timer callback executes, and if we reach this callback, then we know that the button has been long-pressed.

**Comment by bnoordhuis** [comment]
You're accessing the event loop from two threads. That's undefined behavior.
